eslint no-useless-escape

A recent release of eslint added `no-useless-escape` to the recommended ruleset. I reduced the severity of it to a warning for now.

TODO:
* [ ] go through the list of warnings and fix them.
* [ ] remove the rule from .eslint.json

````
[matt@imac27] ~/git/haraka $ npm run lint

> Haraka@2.8.14 lint /Users/matt/git/haraka
> node node_modules/.bin/eslint *.js outbound/*.js plugins/*.js plugins/*/*.js tests/*.js tests/*/*.js bin/haraka bin/spf bin/dkimverify


/Users/matt/git/haraka/configfile.js
   16:32  warning  Unnecessary escape character: \.  no-useless-escape
   16:37  warning  Unnecessary escape character: \/  no-useless-escape
   16:39  warning  Unnecessary escape character: \[  no-useless-escape
  444:31  warning  Unnecessary escape character: \.  no-useless-escape
  454:34  warning  Unnecessary escape character: \-  no-useless-escape
  455:34  warning  Unnecessary escape character: \-  no-useless-escape

/Users/matt/git/haraka/mailbody.js
  272:41  warning  Unnecessary escape character: \"  no-useless-escape
  272:55  warning  Unnecessary escape character: \-  no-useless-escape
  272:63  warning  Unnecessary escape character: \"  no-useless-escape

/Users/matt/git/haraka/mailheader.js
  182:52  warning  Unnecessary escape character: \-  no-useless-escape

/Users/matt/git/haraka/plugins.js
  82:17  warning  Unnecessary escape character: \-  no-useless-escape
  82:25  warning  Unnecessary escape character: \-  no-useless-escape
  83:37  warning  Unnecessary escape character: \-  no-useless-escape
  83:45  warning  Unnecessary escape character: \-  no-useless-escape

/Users/matt/git/haraka/rfc1869.js
  23:47  warning  Unnecessary escape character: \-  no-useless-escape
  57:25  warning  Unnecessary escape character: \@  no-useless-escape
  75:25  warning  Unnecessary escape character: \@  no-useless-escape
  81:25  warning  Unnecessary escape character: \@  no-useless-escape
  88:34  warning  Unnecessary escape character: \@  no-useless-escape

/Users/matt/git/haraka/spf.js
  218:81  warning  Unnecessary escape character: \/  no-useless-escape
  219:48  warning  Unnecessary escape character: \/  no-useless-escape
  233:44  warning  Unnecessary escape character: \/  no-useless-escape
  401:29  warning  Unnecessary escape character: \/  no-useless-escape
  464:29  warning  Unnecessary escape character: \/  no-useless-escape
  557:29  warning  Unnecessary escape character: \/  no-useless-escape
  602:61  warning  Unnecessary escape character: \.  no-useless-escape
  637:24  warning  Unnecessary escape character: \/  no-useless-escape

/Users/matt/git/haraka/tls_socket.js
  159:35   warning  Unnecessary escape character: \/  no-useless-escape
  194:24   warning  Unnecessary escape character: \-  no-useless-escape
  194:32   warning  Unnecessary escape character: \-  no-useless-escape
  194:64   warning  Unnecessary escape character: \-  no-useless-escape
  194:70   warning  Unnecessary escape character: \-  no-useless-escape
  194:75   warning  Unnecessary escape character: \-  no-useless-escape
  194:105  warning  Unnecessary escape character: \-  no-useless-escape

/Users/matt/git/haraka/plugins/attachment.js
  272:32  warning  Unnecessary escape character: \/  no-useless-escape
  281:41  warning  Unnecessary escape character: \.  no-useless-escape
  436:33  warning  Unnecessary escape character: \/  no-useless-escape
  444:54  warning  Unnecessary escape character: \/  no-useless-escape

/Users/matt/git/haraka/plugins/greylist.js
  328:45  warning  Unnecessary escape character: \-  no-useless-escape

/Users/matt/git/haraka/plugins/prevent_credential_leaks.js
  5:26  warning  Unnecessary escape character: \-  no-useless-escape
  5:28  warning  Unnecessary escape character: \[  no-useless-escape
  5:32  warning  Unnecessary escape character: \/  no-useless-escape
  5:34  warning  Unnecessary escape character: \{  no-useless-escape
  5:36  warning  Unnecessary escape character: \}  no-useless-escape
  5:38  warning  Unnecessary escape character: \(  no-useless-escape
  5:40  warning  Unnecessary escape character: \)  no-useless-escape
  5:42  warning  Unnecessary escape character: \*  no-useless-escape
  5:44  warning  Unnecessary escape character: \+  no-useless-escape
  5:46  warning  Unnecessary escape character: \?  no-useless-escape
  5:48  warning  Unnecessary escape character: \.  no-useless-escape
  5:52  warning  Unnecessary escape character: \^  no-useless-escape
  5:54  warning  Unnecessary escape character: \$  no-useless-escape
  5:56  warning  Unnecessary escape character: \|  no-useless-escape

/Users/matt/git/haraka/plugins/auth/auth_vpopmaild.js
  144:19  warning  Unnecessary escape character: \-  no-useless-escape

/Users/matt/git/haraka/bin/haraka
  295:38  warning  Unnecessary escape character: \%  no-useless-escape
  295:45  warning  Unnecessary escape character: \%  no-useless-escape

/Users/matt/git/haraka/bin/spf
  28:15  warning  Unnecessary escape character: \@  no-useless-escape

✖ 57 problems (0 errors, 57 warnings)
````

--- a/Changes.md
+++ b/Changes.md
@@ -2,6 +2,7 @@
 ## 2.8.16 - Mmm dd, 2017
 
 * Changes
+    * lint: remove useless escapes #2117
     * rspamd: repackaged as NPM module #2106
     * dsn: repackaged as NPM module haraka-dsn #2105
     * outbound: add results when queueing #2103

--- a/configfile.js
+++ b/configfile.js
@@ -13,7 +13,7 @@ setImmediate(function () {
 // for "ini" type files
 const regex = exports.regex = {
     section:        /^\s*\[\s*([^\]]*?)\s*\]\s*$/,
-    param:          /^\s*([\w@:\._\-\/\[\]]+)\s*(?:=\s*(.*?)\s*)?$/,
+    param:          /^\s*([\w@:._\-/[\]]+)\s*(?:=\s*(.*?)\s*)?$/,
     comment:        /^\s*[;#].*$/,
     line:           /^\s*(.*?)\s*$/,
     blank:          /^\s*$/,
@@ -441,7 +441,7 @@ cfreader.init_booleans = function (options, result) {
     // Initialize any booleans
     if (options && Array.isArray(options.booleans)) {
         for (let i=0; i<options.booleans.length; i++) {
-            const m = /^(?:([^\. ]+)\.)?(.+)/.exec(options.booleans[i]);
+            const m = /^(?:([^. ]+)\.)?(.+)/.exec(options.booleans[i]);
             if (!m) continue;
 
             let section = m[1] || 'main';
@@ -451,8 +451,8 @@ cfreader.init_booleans = function (options, result) {
                 :     key[0] === '+' ? true
                     :     false;
 
-            if (section.match(/^(\-|\+)/)) section = section.substr(1);
-            if (    key.match(/^(\-|\+)/)) key     =     key.substr(1);
+            if (section.match(/^(-|\+)/)) section = section.substr(1);
+            if (    key.match(/^(-|\+)/)) key     =     key.substr(1);
 
             // so the boolean detection in the next section will match
             if (options.booleans.indexOf(section+'.'+key) === -1) {

--- a/mailbody.js
+++ b/mailbody.js
@@ -269,7 +269,7 @@ Body.prototype.parse_end = function (line) {
     const ct  = this.header.get_decoded('content-type') || 'text/plain';
     let enc = 'UTF-8';
     let pre_enc = '';
-    const matches = /\bcharset\s*=\s*(?:\"|3D|')?([\w_\-]*)(?:\"|3D|')?/.exec(ct);
+    const matches = /\bcharset\s*=\s*(?:"|3D|')?([\w_-]*)(?:"|3D|')?/.exec(ct);
     if (matches) {
         pre_enc = (matches[1]).trim();
         if (pre_enc.length > 0) {

--- a/plugins.js
+++ b/plugins.js
@@ -79,8 +79,8 @@ Plugin.prototype._get_plugin_path = function () {
 
     plugin.hasPackageJson = false;
     let name = plugin.name;
-    if (/^haraka\-plugin\-/.test(name)) {
-        name = name.replace(/^haraka\-plugin\-/, '');
+    if (/^haraka-plugin-/.test(name)) {
+        name = name.replace(/^haraka-plugin-/, '');
     }
 
     let paths = [];

--- a/plugins/attachment.js
+++ b/plugins/attachment.js
@@ -269,7 +269,7 @@ exports.start_attachment = function (connection, ctype, filename, body, stream) 
 
     // Parse Content-Type
     let ct;
-    if ((ct = ctype.match(/^([^\/]+\/[^;\r\n ]+)/)) && ct[1]) {
+    if ((ct = ctype.match(/^([^/]+\/[^;\r\n ]+)/)) && ct[1]) {
         connection.logdebug(plugin, 'found content type: ' + ct[1]);
         txn.notes.attachment_ctypes.push(ct[1]);
     }
@@ -278,7 +278,7 @@ exports.start_attachment = function (connection, ctype, filename, body, stream) 
     let ext;
     let fileext = '.unknown';
     if (filename) {
-        if ((ext = filename.match(/(\.[^\. ]+)$/)) && ext[1]) {
+        if ((ext = filename.match(/(\.[^. ]+)$/)) && ext[1]) {
             fileext = ext[1].toLowerCase();
         }
         txn.notes.attachment_files.push(filename);
@@ -433,7 +433,7 @@ exports.check_attachments = function (next, connection) {
     // Add in any content type from message body
     const body = txn.body;
     let body_ct;
-    if (body && (body_ct = /^([^\/]+\/[^;\r\n ]+)/.exec(body.header.get('content-type')))) {
+    if (body && (body_ct = /^([^/]+\/[^;\r\n ]+)/.exec(body.header.get('content-type')))) {
         connection.logdebug(this, 'found content type: ' + body_ct[1]);
         ctypes.push(body_ct[1]);
     }
@@ -441,7 +441,7 @@ exports.check_attachments = function (next, connection) {
     if (body && body.children) {
         for (let c=0; c<body.children.length; c++) {
             let child_ct;
-            if (body.children[c] && (child_ct = /^([^\/]+\/[^;\r\n ]+)/.exec(body.children[c].header.get('content-type')))) {
+            if (body.children[c] && (child_ct = /^([^/]+\/[^;\r\n ]+)/.exec(body.children[c].header.get('content-type')))) {
                 connection.logdebug(this, 'found content type: ' + child_ct[1]);
                 ctypes.push(child_ct[1]);
             }

--- a/plugins/prevent_credential_leaks.js
+++ b/plugins/prevent_credential_leaks.js
@@ -2,14 +2,14 @@
 // This is a simple, primitive form of anti-phishing.
 
 function escapeRegExp (str) {
-    return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+    return str.replace(/[-[\]/{}()*+?.\\^$|]/g, "\\$&");
 }
 
 exports.hook_data = function (next, connection) {
     if (connection.notes.auth_user && connection.notes.auth_passwd) {
         connection.transaction.parse_body = true;
     }
-    return next();
+    next();
 };
 
 exports.hook_data_post = function (next, connection) {
@@ -38,7 +38,7 @@ exports.hook_data_post = function (next, connection) {
         return next(DENY, "Credential leak detected: never give out your username/password to anyone!");
     }
 
-    return next();
+    next();
 };
 
 function look_for_credentials (user_regexp, passwd_regexp, body) {

--- a/rfc1869.js
+++ b/rfc1869.js
@@ -1,4 +1,4 @@
-"use strict";
+'use strict';
 // RFC 1869 command parser
 
 // 6.  MAIL FROM and RCPT TO Parameters
@@ -20,7 +20,7 @@
 //                        ("RCPT TO:" forward-path)
 
 /* eslint no-control-regex: 0 */
-const chew_regexp = /\s+([A-Za-z0-9][A-Za-z0-9\-]*(?:=[^= \x00-\x1f]+)?)$/;
+const chew_regexp = /\s+([A-Za-z0-9][A-Za-z0-9-]*(?:=[^= \x00-\x1f]+)?)$/;
 
 exports.parse = function (type, line, strict) {
     let params = [];
@@ -54,7 +54,7 @@ exports.parse = function (type, line, strict) {
     if (line.length) {
         // parameter syntax error, i.e. not all of the arguments were
         // stripped by the while() loop:
-        if (line.match(/\@.*\s/)) {
+        if (line.match(/@.*\s/)) {
             throw new Error("Syntax error in parameters (" + line + ")");
         }
 
@@ -72,20 +72,20 @@ exports.parse = function (type, line, strict) {
         if (!line.length) {
             return ["<>"]; // 'MAIL FROM:' --> 'MAIL FROM:<>'
         }
-        if (line.match(/\@.*\s/)) {
+        if (line.match(/@.*\s/)) {
             throw new Error("Syntax error in parameters");
         }
     }
     else {
         // console.log("Looking at " + line);
-        if (line.match(/\@.*\s/)) {
+        if (line.match(/@.*\s/)) {
             throw new Error("Syntax error in parameters");
         }
         else {
             if (line.match(/\s/)) {
                 throw new Error("Syntax error in parameters");
             }
-            else if (line.match(/\@/)) {
+            else if (line.match(/@/)) {
                 if (!line.match(/^<.*>$/)) {
                     line = '<' + line + '>';
                 }

--- a/spf.js
+++ b/spf.js
@@ -215,8 +215,8 @@ SPF.prototype.check_host = function (ip, domain, mail_from, cb) {
         const mech_array = [];
         const mod_array = [];
         const mech_regexp1 = /^([-+~?])?(all|a|mx|ptr)$/;
-        const mech_regexp2 = /^([-+~?])?(a|mx|ptr|ip4|ip6|include|exists)((?::[^\/ ]+(?:\/\d+(?:\/\/\d+)?)?)|\/\d+(?:\/\/\d+)?)$/;
-        const mod_regexp = /^([^ =]+)=([a-z0-9:\/._-]+)$/;
+        const mech_regexp2 = /^([-+~?])?(a|mx|ptr|ip4|ip6|include|exists)((?::[^/ ]+(?:\/\d+(?:\/\/\d+)?)?)|\/\d+(?:\/\/\d+)?)$/;
+        const mod_regexp = /^([^ =]+)=([a-z0-9:/._-]+)$/;
         const split = spf_record.split(' ');
         for (let i=1; i<split.length; i++) {
             // Skip blanks
@@ -230,7 +230,7 @@ SPF.prototype.check_host = function (ip, domain, mail_from, cb) {
                 self.log_debug('found mechanism: ' + match);
                 // Validate IP addresses
                 if (match[2] === 'ip4' || match[2] === 'ip6') {
-                    const ip_split = /^:([^\/ ]+)(?:\/([^ ]+))?$/.exec(match[3]);
+                    const ip_split = /^:([^/ ]+)(?:\/([^ ]+))?$/.exec(match[3]);
                     // Make sure the IP address is valid
                     if (!ip_split || (ip_split && !ipaddr.isValid(ip_split[1]))) {
                         self.log_debug('invalid IP address: ' + ip_split[1]);
@@ -398,7 +398,7 @@ SPF.prototype.mech_a = function (qualifier, args, cb) {
     }
     let dm;
     let domain = this.domain;
-    if (args && (dm = /^:([^\/ ]+)/.exec(args))) {
+    if (args && (dm = /^:([^/ ]+)/.exec(args))) {
         domain = dm[1];
     }
     // Calculate with IP method to use
@@ -461,7 +461,7 @@ SPF.prototype.mech_mx = function (qualifier, args, cb) {
     }
     let dm;
     let domain = this.domain;
-    if (args && (dm = /^:([^\/ ]+)/.exec(args))) {
+    if (args && (dm = /^:([^/ ]+)/.exec(args))) {
         domain = dm[1];
     }
     // Fetch the MX records for the specified domain
@@ -554,7 +554,7 @@ SPF.prototype.mech_ptr = function (qualifier, args, cb) {
     this.count++;
     let dm;
     let domain = this.domain;
-    if (args && (dm = /^:([^\/ ]+)/.exec(args))) {
+    if (args && (dm = /^:([^/ ]+)/.exec(args))) {
         domain = dm[1];
     }
     // First do a PTR lookup for the connecting IP
@@ -599,7 +599,7 @@ SPF.prototype.mech_ptr = function (qualifier, args, cb) {
                         // Catch bogus PTR matches e.g. ptr:*.bahnhof.se (should be ptr:bahnhof.se)
                         // These will cause a regexp error, so we can catch them.
                         try {
-                            re = new RegExp(domain.replace('\.','\\.') + '$', 'i');
+                            re = new RegExp(domain.replace('.','\\.') + '$', 'i');
                         }
                         catch (e) {
                             self.log_debug(
@@ -634,7 +634,7 @@ SPF.prototype.mech_ptr = function (qualifier, args, cb) {
 
 SPF.prototype.mech_ip = function (qualifier, args, cb) {
     const cidr = args.substr(1);
-    const match = /^([^\/ ]+)(?:\/(\d+))?$/.exec(cidr);
+    const match = /^([^/ ]+)(?:\/(\d+))?$/.exec(cidr);
     if (!match) { return cb(null, this.SPF_NONE); }
 
     // match[1] == ip

